File: adapters/android/OeapActionHubClient.kt
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

/**
 * Dependency-free Android client for OEAP AI Action Hub.
 *
 * Configure after the app has obtained an OEAP/enterprise session token.
 * Call from a background coroutine/thread because network I/O is blocking.
 */
object OeapActionHubClient {
    data class Configuration(
        val baseUrl: String,
        val bearerToken: String,
        val organizationId: String? = null,
        val memberId: String? = null,
        val connectTimeoutMs: Int = 15_000,
        val readTimeoutMs: Int = 120_000
    )

    @Volatile
    private var configuration: Configuration? = null

    fun configure(configuration: Configuration) {
        this.configuration = configuration.copy(
            baseUrl = configuration.baseUrl.trimEnd('/')
        )
    }

    /**
     * Execute a Universal Action and return the raw Action Hub JSON response.
     * R2/R3 actions normally return an `approval_required` body with HTTP 202.
     */
    fun execute(
        actionId: String,
        payloadJson: String,
        preferredProvider: String? = null
    ): String {
        val config = configuration
            ?: error("OeapActionHubClient is not configured")

        val input = try {
            JSONObject(payloadJson)
        } catch (error: Exception) {
            throw IllegalArgumentException(
                "Action payload must be a JSON object",
                error
            )
        }

        val encodedAction = URLEncoder.encode(
            actionId,
            StandardCharsets.UTF_8.name()
        ).replace("+", "%20")
        val endpoint = URL(
            "${config.baseUrl}/api/action-hub/actions/$encodedAction/execute"
        )
        val body = JSONObject().put("input", input)
        if (!preferredProvider.isNullOrBlank()) {
            body.put("preferredProvider", preferredProvider)
        }

        val connection = endpoint.openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = config.connectTimeoutMs
            connection.readTimeout = config.readTimeoutMs
            connection.doOutput = true
            connection.setRequestProperty(
                "Content-Type",
                "application/json"
            )
            connection.setRequestProperty(
                "Authorization",
                "Bearer ${config.bearerToken}"
            )
            config.organizationId?.let {
                connection.setRequestProperty("X-OEAP-Org", it)
            }
            config.memberId?.let {
                connection.setRequestProperty("X-OEAP-Member", it)
            }

            connection.outputStream.use { output ->
                output.write(
                    body.toString().toByteArray(StandardCharsets.UTF_8)
                )
            }

            val status = connection.responseCode
            val stream = if (status in 200..299) {
                connection.inputStream
            } else {
                connection.errorStream
            }
            val response = stream?.bufferedReader(
                StandardCharsets.UTF_8
            )?.use { it.readText() }.orEmpty()

            if (status !in 200..299) {
                throw IllegalStateException(
                    "Action Hub HTTP $status: $response"
                )
            }

            return response
        } finally {
            connection.disconnect()
        }
    }
}

File: adapters/apple/OEAPActionHubClient.swift
import Foundation

/// Small dependency-free client for calling OEAP AI Action Hub from an Apple app.
/// Configure it once after your app has obtained an OEAP/enterprise session token.
public final class OEAPActionHubClient: @unchecked Sendable {
    public static let shared = OEAPActionHubClient()

    public struct Configuration: Sendable {
        public let baseURL: URL
        public let bearerToken: String
        public let organizationID: String?
        public let memberID: String?

        public init(
            baseURL: URL,
            bearerToken: String,
            organizationID: String? = nil,
            memberID: String? = nil
        ) {
            self.baseURL = baseURL
            self.bearerToken = bearerToken
            self.organizationID = organizationID
            self.memberID = memberID
        }
    }

    public enum ClientError: LocalizedError {
        case notConfigured
        case invalidPayload
        case invalidResponse
        case http(status: Int, body: String)

        public var errorDescription: String? {
            switch self {
            case .notConfigured:
                return "OEAPActionHubClient is not configured"
            case .invalidPayload:
                return "Action payload is not valid JSON"
            case .invalidResponse:
                return "Action Hub returned an invalid response"
            case let .http(status, body):
                return "Action Hub HTTP \(status): \(body)"
            }
        }
    }

    private let lock = NSLock()
    private var configuration: Configuration?
    private var session: URLSession = .shared

    private init() {}

    public func configure(
        _ configuration: Configuration,
        session: URLSession = .shared
    ) {
        lock.lock()
        defer { lock.unlock() }
        self.configuration = configuration
        self.session = session
    }

    /// Executes a Universal Action and returns the raw Action Hub JSON response.
    /// R2/R3 actions normally return an `approval_required` response (HTTP 202).
    public func execute(
        actionID: String,
        payloadJSON: String,
        preferredProvider: String? = nil
    ) async throws -> String {
        let state = try configuredState()
        guard let payloadData = payloadJSON.data(using: .utf8),
              let payload = try? JSONSerialization.jsonObject(with: payloadData) else {
            throw ClientError.invalidPayload
        }

        let encodedID = actionID.addingPercentEncoding(
            withAllowedCharacters: .urlPathAllowed
        ) ?? actionID
        let endpoint = state.configuration.baseURL
            .appendingPathComponent("api")
            .appendingPathComponent("action-hub")
            .appendingPathComponent("actions")
            .appendingPathComponent(encodedID)
            .appendingPathComponent("execute")

        var body: [String: Any] = ["input": payload]
        if let preferredProvider, !preferredProvider.isEmpty {
            body["preferredProvider"] = preferredProvider
        }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(
            "Bearer \(state.configuration.bearerToken)",
            forHTTPHeaderField: "Authorization"
        )
        if let organizationID = state.configuration.organizationID {
            request.setValue(organizationID, forHTTPHeaderField: "X-OEAP-Org")
        }
        if let memberID = state.configuration.memberID {
            request.setValue(memberID, forHTTPHeaderField: "X-OEAP-Member")
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await state.session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ClientError.invalidResponse
        }

        let text = String(data: data, encoding: .utf8) ?? ""
        guard (200..<300).contains(http.statusCode) else {
            throw ClientError.http(status: http.statusCode, body: text)
        }
        return text
    }

    private func configuredState() throws -> (
        configuration: Configuration,
        session: URLSession
    ) {
        lock.lock()
        defer { lock.unlock() }
        guard let configuration else {
            throw ClientError.notConfigured
        }
        return (configuration, session)
    }
}
